docs/proposals/resource-catalog.md

```diff
@@ -0,0 +1,222 @@
+
+
+We can reshape the catalog resource to more easily express them, secure them, and claim them.
+
+# Goals
+ 1.  cluster-admin should control which catalog exist.
+ 2.  cluster-admin should be able/disable per-catalog exposure on a per-project granularity.
+ 3.  project-admin should be able to delegate per-catalog exposure.
+ 4.  project-editor should be able to claim an entry from the catalog.
+ 5.  ResourceCatalogClaims should be sync-able, but not require syncing.
+ 6.  project-admin should be able to control the shape of claim realization and prevent unwanted injection.
+ 7.  ResourceCatalogEntries should not be restricted to specific claim realization mechanism.
+
+To achieve the ACL related goals, we can create namespaced resources that are selected by label at a cluster level.
+
+
+# API Objects
+
+## ResourceCatalog
+ResourceCatalog is a cluster-scoped resource.  It has a cluster-admin controlled spec that is used to drive a controller.
+The spec includes a label selector that is used in a list/watch to filter all ResourceCatalogEntries in the cluster.
+The controller will find matching entries and copy a reference into the ResourceCatalog.Status field.  The controller
+should have a plug-point for deciding if a given entry is accepted into a given catalog: `Accept(catalogName, entryNamespace, entryName)`
+is probably sufficient.
+
+```go
+// ResourceCatalog is a cluster-scoped resource with a specified spec and a controller maintained status
+type ResourceCatalog struct {
+	unversioned.TypeMeta
+	api.ObjectMeta
+
+	Spec   ResourceCatalogSpec
+	Status ResourceCatalogStatus
+}
+
+type ResourceCatalogSpec struct {
+	// Selector is the selector to use when gathering the namespace scope `ResourceCatalogEntries` for
+	// the `Status`
+	Selector map[string]string
+}
+
+type ResourceCatalogStatus struct {
+	// Entries is dynamically retrieved by selecting all matching ResourceCatalogEntries from all namespaces
+	// This can be done by a controller using a list/watch on all ResourceCatalogEntries based on the spec.selector.
+	Entries []ResourceCatalogInfo
```

Comment:
Ultimately project lacks sane watch, but we haven't regretted the design.
Good enough watch is sufficient.

I would be extremely surprised to catch us doing a watch on catalog entries
except as an admin monitoring tool (which should have access to everything
in the catalog).

On Tue, May 31, 2016 at 1:43 PM, David Eads <notifications@github.com>
wrote:

> In docs/proposals/resource-catalog.md
> <https://github.com/ncdc/kubernetes/pull/8#discussion_r65229484>:
>
> > +	api.ObjectMeta
> > +
> > +	Spec   ResourceCatalogSpec
> > +	Status ResourceCatalogStatus
> > +}
> > +
> > +type ResourceCatalogSpec struct {
> > +	// Selector is the selector to use when gathering the namespace scope `ResourceCatalogEntries` for
> > +	// the `Status`
> > +	Selector map[string]string
> > +}
> > +
> > +type ResourceCatalogStatus struct {
> > +	// Entries is dynamically retrieved by selecting all matching ResourceCatalogEntries from all namespaces
> > +	// This can be done by a controller using a list/watch on all ResourceCatalogEntries based on the spec.selector.
> > +	Entries []ResourceCatalogInfo
>
> I'd rather see the list of entries as a virtual resource and not part of
> the catalog
>
> Ok, but that will cost you a sane watch.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ncdc/kubernetes/pull/8/files/a2acd970c6316ce196c7bb3f24fb57e1bbef647e#r65229484>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p5vOGdhmZOygeFyhhPeey2vDiworks5qHHNRgaJpZM4IO9RU>
> .
>